README.md
# iconfount

`iconfount`是一个离线icon font生成工具，支持从多个svg文件生成一套字体以及对应的样式文件。

[![NPM](https://nodei.co/npm/iconfount.png?downloads=true)](https://nodei.co/npm/iconfount/)

## 字体文件兼容性

1. 桌面端浏览器：IE6+, Safari, FF, Chrome
2. 移动端浏览器：绝大部分系统都没有问题，部分*非常老*的安卓系统或者某些古董浏览器可能会有问题。不支持Windows Mobile IE9。

## 安装

`ynpm install -g iconfount` or `yarn global add iconfount`.

## 第三方依赖

这些依赖是可选的，用到特定选项时才需要安装。

1. ttfautohint: `brew install ttfautohint`, `hinting`是`true`的时候需要安装
2. fontforge: 参见[安装fontforge](https://github.com/youzan/iconfount/blob/master/doc/fontforge.md), 图标`correct_contour_direction`为`true`时需要安装

## 使用

### CLI

```text
Usage: iconfount --config file [options]

Options:
  --help        print this message and exit  [boolean]
  -c, --config  config file to read  [required]
  -t, --test    test configuration and exit  [boolean]
```

### API

#### iconfount(config): buildConfig

* `config` - config file to read  [required]
* `buildConfig` - full config for generate, include svg infos  [promise]

``` js
const iconfount = require('iconfount')

iconfount('file')
```

## 输出文件

```text
├── LICENSE.txt                       版权信息
├── README.txt                        说明文件
├── codes.json                        字符的详细定义，可用于外部工具生成样式代码
├── config.json                       调试用，请忽略
├── css
│   ├── animation.css                 调试用，动画
│   ├── zenticons-codes.css           只包含字符codepoint的css文件
│   ├── zenticons-embedded.css        区别在于WOFF内联了，规避CORS时有用
│   ├── zenticons-ie7-codes.css       IE7专用
│   ├── zenticons-ie7.css             IE7专用
│   └── zenticons.css                 完整的样式，包括font-face以及各个字符的样式
├── demo.html                         预览文件
└── font
    ├── zenticons-db0a845be8.eot      IE专用
    ├── zenticons-db0a845be8.svg      老的Safari专用
    ├── zenticons-db0a845be8.ttf
    ├── zenticons-db0a845be8.woff
    └── zenticons-db0a845be8.woff2
```

css目录下已经提供了两种样式选择：完整的和只包含字符定义的css。如果仍不能满足需求，可以使用`codes.json`
文件自己生成样式。


## 示例

`sample`目录下提供了一个示例配置。

## 配置文件

支持`json`或者`js`文件，`js`文件的话直接`export`一个包含配置信息的对象。

`sample/`目录下有一个示例配置文件，仅供参考。

`doc/`目录下有一些文档，对详细理解一些配置有帮助。

### 参数

#### name

type: `string`

required: `false` 

default: `'iconfount'`

字体名字，最好英文吧

#### output

type: `string`

required: `true`

输出字体、样式以及示例文件的目录，可以是相对路径或者绝对路径

#### hinting

type: `boolean`

required: `false`

default: `false`

是否使用`ttfautohint`对字体文件做hint，需要安装`ttfautohint`。
建议开启。

#### units_per_em

type: `number`

required: `false`

default: `1000`

简单理解为单个svg图标的画板大小，[detail](https://www.w3.org/TR/SVG/fonts.html#FontFaceElementUnitsPerEmAttribute)
不清楚的话不要设置。

#### ascent

type: `number`

required: `false`

default: `850`

从baseline到最高字符顶部的高度, [detail](https://www.w3.org/TR/SVG/fonts.html#FontFaceElementAscentAttribute)
不清楚的话不要设置。

#### weight

type: `number`

required: `false`

default: `400`

字体的粗细，和`font-weight`属性一致

#### start_codepoint

type: `number`

required: `false`

default: `59392`, 即`0xE800`。

最小`0xE000`(`57344`)，最大`0xF8FF`(`63743`)，这个区间是
Unicode的[Private Use Area](https://en.wikipedia.org/wiki/Private_Use_Areas#Private_Use_Areas)，
虽然planes 15 and 16也是Private Use Areas，但是我们不支持这两个区间。

#### glyphs_dir

type: `string`

required: `true`

存放svg文件的根目录，**如果是相对路径，必须是相对当前配置文件的路径**。

#### glyphs

type: `Array.<Glyph>`

required: `true`

所有图标的定义，每个图标有如下几个属性.

##### glyphs.keywords

type: `Array.<string>`

required: `false`

default: `[]`

图标的一些tag，主要是描述图标的作用.

##### glyphs.src

type: `string`

required: `true`

图标相对`glyphs_dir`的路径。

##### glyphs.css

type: `string`

required: `true`

图标的CSS名字，图标最终的class是`${meta.css_prefix_text}${css}`，如果`meta.css_use_suffix`
是`true`，图标最终的class是`${css}${meta.css_prefix_text}`。

##### glyphs.correct_contour_direction

type: `boolean`

required: `false`

default: `false`

是否使用`fontforge`对字符做轮廓方向修正，需要安装`fontforge`。
视情况开启，如果生成的字符填充位置不正常，再考虑开启。不过做好是让设计修改原图的路径方向，具体请看
`doc/contour-direction.md`

#### meta

type: `Object`

required: `false`

一些元数据

##### meta.author

type: `string`

required: `false`

default: `''`

字体的作者

##### meta.license

type: `string`

required: `false`

default: `''`

字体的许可协议

##### meta.license_url

type: `string`

required: `false`

default: `''`

字体许可协议链接

##### meta.homepage

type: `string`

required: `false`

default: `''`

字体主页

##### meta.css_prefix_text

type: `boolean`

required: `false`

default: `'icon-'`

图标css前缀。

#### meta.css_use_suffix

type: `boolean`

required: `false`

default: `false`

如果为`true`, `meta.css_prefix_text`将被用作css后缀。

##### meta.columns

type: `number`

required: `false`

default: `4`

示例页面中每行的列数。

##### meta.filename_hash

type: `boolean`

required: `false`

default: `true`

是否在文件名中加入随机字符串，确保每次生成的文件名都不同。

## License

iconfount's code (all files) is distributed under MIT license. See
[LICENSE](https://github.com/youzan/iconfount/blob/master/LICENSE) file for details.

Many thanks to the [fontello](https://github.com/fontello/fontello) project.


CHANGELOG.md
# Change logs

## 2.3.0 (2020-04-12)

* 升级包依赖至最新
* 使用 `pugjs` 替换 `jade`
* 微调 demo 页面样式

## 2.2.0 (2019-11-07)

* 升级依赖保本版
* 增加 Node.js `API` 的使用文档

## 2.1.1 (2018-10-26)

* 修复 `svgo` 代码问题

## 2.1.0 (2018-10-25)

* 关闭 `svgo` 中的 `convertPathData` 优化，会引起边粗细不一致的问题

## 2.0.0 (2017.02.08)

* 项目开源，重命名为iconfount(因为npm包名被占用)

## 1.0.9 (2017.01.16)

* output现在是相对配置文件的路径，和glyphDir保持一致

## 1.0.8 (2016.12.22)

* 调整外部工具检测的报错流程

## 1.0.7 (2016.12.21)

* 调用外部命令时先判断命令是否存在，ttfautohint貌似会fail silently
* 在部分文件里添加了fontello的license



src/index.js
/**
 * Modified from fontello project: https://github.com/fontello/fontello
 *
 * (The MIT License)
 *
 * Copyright (C) 2011 by Vitaly Puzrin
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

"use strict";

var _ = require("lodash");
var svgpath = require("svgpath");
var path = require("path");
var fs = require("fs");
var rimraf = require("rimraf");
var mkdirp = require("mkdirp");
var svg2ttf = require("svg2ttf");
var ttf2eot = require("ttf2eot");
var ttf2woff = require("ttf2woff");
var ttf2woff2 = require("ttf2woff2");
var child_process = require("child_process");
var pug = require("pug");
var b64 = require("base64-js");
var crypto = require("crypto");
var { optimize, extendDefaultPlugins } = require("svgo");

var loadSvg = require("./svg");
var log = require("./log");

var ERROR = {
  ttfautohint: "ttfautohint not installed",
  fontforge: "fontforge python extension not installed",
};

var TEMPLATES_DIR = path.join(__dirname, "../templates");
var SVG_FONT_TEMPLATE = _.template(
  fs.readFileSync(path.join(TEMPLATES_DIR, "font/svg.tpl"), "utf8")
);
var TEMPLATES = _.reduce(
  {
    "demo.pug": "demo.html",
    "css/css.pug": "css/${FONTNAME}.css",
    "css/css-ie7.pug": "css/${FONTNAME}-ie7.css",
    "css/css-codes.pug": "css/${FONTNAME}-codes.css",
    "css/css-ie7-codes.pug": "css/${FONTNAME}-ie7-codes.css",
    "css/css-embedded.pug": "css/${FONTNAME}-embedded.css",
    "LICENSE.pug": "LICENSE.txt",
    "css/animation.css": "css/animation.css",
    "README.txt": "README.txt",
  },
  function (templates, outputName, inputName) {
    var inputFile = path.join(TEMPLATES_DIR, inputName);
    var inputData = fs.readFileSync(inputFile, "utf8");
    var outputData;

    switch (path.extname(inputName)) {
      case ".pug": // Pugjs template.
        outputData = pug.compile(inputData, {
          pretty: true,
          filename: inputFile,
        });
        break;

      case ".tpl": // Lodash template.
        outputData = _.template(inputData);
        break;

      default:
        // Static file - just do a copy.
        outputData = _.constant(inputData);
        break;
    }

    templates[outputName] = outputData;
    return templates;
  },
  {}
);

function getRandomString() {
  var hash = crypto.createHash("sha256");
  var randomNumber = Date.now() + _.random(0, 1, true) * 1000000;
  hash.update(randomNumber + "");
  return hash.digest("hex").slice(-10);
}

/**
 * 生成glyphs的json信息，可以给外部工具用来生成代码
 * @return {[type]} [description]
 */
function generateGlyphsJson(buildConfig) {
  var prefix = buildConfig.meta.css_prefix_text;
  var useSuffix = buildConfig.meta.css_use_suffix;
  var glyphs = buildConfig.glyphs;

  if (_.isEmpty(glyphs)) {
    return [];
  }

  return glyphs.map(function (g) {
    var css = useSuffix ? g.css + prefix : prefix + g.css;
    var hex = "0x" + g.code.toString(16);
    return {
      name: g.css,
      css: css,
      hexCodepoint: hex,
      codepoint: g.code,
      keywords: g.keywords,
    };
  });
}

function collectGlyphsInfo(config) {
  var scale = config.units_per_em / 1000;
  var badGlyphs = [];
  var result = _.map(config.glyphs, function (glyph) {
    var sp = svgpath(glyph.svg.path)
      .scale(scale, -scale)
      .translate(0, config.ascent)
      .abs()
      .round(0)
      .rel();

    // 把需要调整路径方向的字符保存起来，后面会交给fontforge处理
    if (glyph.correct_contour_direction) {
      badGlyphs.push(glyph.code);
    }

    return _.assign({}, _.pick(glyph, ["src", "code", "css", "keywords"]), {
      width: +(glyph.svg.width * scale).toFixed(1),
      d: sp.toString(),
      segments: sp.segments.length,
    });
  });

  // Sort result by original codes.
  result.sort((a, b) => a.code - b.code);

  return {
    glyphs: result,
    badGlyphs: badGlyphs,
  };
}

function fontConfig(config) {
  config.hinting = !!config.hinting;
  config.correct_contour_direction = !!config.correct_contour_direction;
  config.units_per_em = +config.units_per_em || 1000;
  config.ascent = +config.ascent || 850;
  config.weight = config.weight || 400;
  config.copyright =
    config.copyright ||
    "Copyright (C) " + new Date().getFullYear() + " by original authors";

  var fontname;
  if (!_.isEmpty(config.name)) {
    fontname = String(config.name).replace(/[^a-z0-9\-_]+/g, "-");
  } else {
    fontname = "iconfount";
  }

  var glyphsInfo = collectGlyphsInfo(config);
  if (_.isEmpty(glyphsInfo.glyphs)) return null;

  var meta = config.meta || {};
  return {
    font: {
      fontname,
      fullname: fontname,
      // !!! IMPORTANT for IE6-8 !!!
      // due bug, EOT requires `familyname` begins `fullname`
      // https://github.com/fontello/fontello/issues/73?source=cc#issuecomment-7791793
      familyname: fontname,
      copyright: config.copyright,
      ascent: config.ascent,
      descent: config.ascent - config.units_per_em,
      weight: config.weight,
    },
    output: path.resolve(config.output),
    hinting: config.hinting,
    meta: _.extend({}, meta, {
      // Used by the demo page
      columns: meta.columns || 4,
      css_prefix_text: meta.css_prefix_text || "icon-",
      css_use_suffix: Boolean(meta.css_use_suffix),

      // append hash to font file name
      filename_hash: Boolean(meta.filename_hash) ? "-" + getRandomString() : "",
    }),
    glyphs: glyphsInfo.glyphs,
    badGlyphs: glyphsInfo.badGlyphs,
  };
}

function generateFonts(buildConfig) {
  var timeStart = Date.now();
  var fontname = buildConfig.font.fontname;
  var files;

  log.info("Start generation:");

  // Collect file paths.
  var fileNameHash = buildConfig.meta.filename_hash;
  var fontNameWithHash = fontname + fileNameHash;
  files = {
    codes: path.join(buildConfig.output, "codes.json"),
    config: path.join(buildConfig.output, "config.json"),
    svg: path.join(buildConfig.output, "font", fontNameWithHash + ".svg"),
    ttf: path.join(buildConfig.output, "font", fontNameWithHash + ".ttf"),
    ttfUnhinted: path.join(
      buildConfig.output,
      "font",
      fontNameWithHash + "-unhinted.ttf"
    ),
    ttfDirectionNotCorrected: path.join(
      buildConfig.output,
      "font",
      fontNameWithHash + "-direction-not-corrected.ttf"
    ),
    eot: path.join(buildConfig.output, "font", fontNameWithHash + ".eot"),
    woff: path.join(buildConfig.output, "font", fontNameWithHash + ".woff"),
    woff2: path.join(buildConfig.output, "font", fontNameWithHash + ".woff2"),
  };

  // Prepare temporary working directory.
  rimraf.sync(buildConfig.output);
  mkdirp.sync(buildConfig.output);
  mkdirp.sync(path.join(buildConfig.output, "font"));
  mkdirp.sync(path.join(buildConfig.output, "css"));

  // Write client config
  var configOutput = JSON.stringify(buildConfig, null, 2);
  fs.writeFileSync(files.config, configOutput, "utf8");
  log.info("write build config");

  // Write codes
  var codesJSON = JSON.stringify(generateGlyphsJson(buildConfig), null, 2);
  fs.writeFileSync(files.codes, codesJSON, "utf8");
  log.info("write codes");

  // Generate initial SVG font.
  var svgOutput = SVG_FONT_TEMPLATE(buildConfig);
  fs.writeFileSync(files.svg, svgOutput, "utf8");
  log.info("write svg font");

  // Convert SVG to TTF
  var ttf = svg2ttf(svgOutput, { copyright: buildConfig.font.copyright });
  fs.writeFileSync(files.ttf, Buffer.from(ttf.buffer));
  log.info("write ttf first-pass");

  // Autohint the resulting TTF.
  // Don't allow hinting if font has "strange" glyphs.
  // That's useless anyway, and can hang ttfautohint < 1.0
  var maxSegments = _.maxBy(buildConfig.glyphs, (glyph) => glyph.segments)
    .segments;
  if (maxSegments <= 500 && buildConfig.hinting) {
    // 貌似未处理的异常被吃掉了，所以先判断命令是否存在
    var ttfautohintInstalled = false;
    try {
      child_process.execFileSync("ttfautohint", ["--version"], {
        cwd: process.cwd(),
      });
      ttfautohintInstalled = true;
    } catch (ex) {}

    if (!ttfautohintInstalled) {
      log.error(ERROR.ttfautohint);
      process.exit(-2);
    }

    log.info("autohint with ttfautohint");

    fs.renameSync(files.ttf, files.ttfUnhinted);
    child_process.execFileSync(
      "ttfautohint",
      [
        "--no-info",
        "--windows-compatibility",
        "--symbol",
        // temporary workaround for #464
        // https://github.com/fontello/fontello/issues/464#issuecomment-202244651
        "--fallback-script=latn",
        files.ttfUnhinted,
        files.ttf,
      ],
      { cwd: process.cwd() }
    );

    fs.unlinkSync(files.ttfUnhinted);
    log.info("write ttf autohint");
  }

  // 主要是处理有些fill-rule设置为evenodd的svg文件，因为ttf支持non-zero的模式，所以会导致
  // 生成的字体文件里有些路径被错误填充了
  // 这个功能依赖fontforge
  if (!_.isEmpty(buildConfig.badGlyphs)) {
    // 貌似未处理的异常被吃掉了，所以先判断命令是否存在
    var fontforgeInstalled = false;
    try {
      child_process.execFileSync(
        path.resolve(__dirname, "../scripts/test-fontforge.py"),
        [],
        { cwd: process.cwd() }
      );
      fontforgeInstalled = true;
    } catch (ex) {}

    if (!fontforgeInstalled) {
      log.error(ERROR.fontforge);
      process.exit(-1);
    }

    log.info("try correct contour direction with fontforge");

    fs.renameSync(files.ttf, files.ttfDirectionNotCorrected);
    child_process.execFileSync(
      path.resolve(__dirname, "../scripts/correct-direction.py"),
      [files.ttfDirectionNotCorrected, files.ttf].concat(buildConfig.badGlyphs),
      { cwd: process.cwd() }
    );
    fs.unlinkSync(files.ttfDirectionNotCorrected);
    log.info("write ttf correct direction");
  }

  // generate other font types
  var ttfOutput = new Uint8Array(fs.readFileSync(files.ttf));
  var eotOutput = ttf2eot(ttfOutput).buffer;
  fs.writeFileSync(files.eot, Buffer.from(eotOutput));
  log.info("write eot file");

  var woffOutput = ttf2woff(ttfOutput).buffer;
  fs.writeFileSync(files.woff, Buffer.from(woffOutput));
  log.info("write woff file");

  var woff2Output = ttf2woff2(ttfOutput).buffer;
  fs.writeFileSync(files.woff2, Buffer.from(woff2Output));
  log.info("write woff2 file");

  // Write template files. (generate dynamic and copy static)
  _.forEach(TEMPLATES, function (templateData, templateName) {
    var outputName = templateName.replace("${FONTNAME}", fontname);
    var outputFile = path.join(buildConfig.output, outputName);
    var outputData = templateData(buildConfig);

    outputData = outputData
      .replace("%WOFF64%", b64.fromByteArray(woffOutput))
      .replace("%TTF64%", b64.fromByteArray(ttfOutput));

    fs.writeFileSync(outputFile, outputData, "utf8");
  });
  log.info("write demo/css files");

  var timeEnd = Date.now();
  log.info("Generated in " + (timeEnd - timeStart) / 1000);
}

module.exports = function build(configFile) {
  var configFilePath = path.resolve(configFile);
  var config = require(configFilePath);
  config.start_codepoint = config.start_codepoint || 0xe800;

  // glyphs dir is relative to config file
  var glyphsDir = config.glyphs_dir || process.cwd();
  if (!path.isAbsolute(glyphsDir)) {
    glyphsDir = path.join(path.dirname(configFilePath), glyphsDir);
  }

  // output dir is relative to config file
  var outputDir = config.output;
  if (!path.isAbsolute(outputDir)) {
    outputDir = path.join(path.dirname(configFilePath), outputDir);
  }
  config.output = outputDir;

  var glyphs = config.glyphs.map(function (glyph) {
    var xml = fs.readFileSync(path.join(glyphsDir, glyph.src), "utf-8");

    const cleanXml = optimize(xml, {
      plugins: extendDefaultPlugins([
        // 这个插件开启之后会导致有些圆一边粗一边细
        {
          name: "convertPathData",
          active: false,
        },
      ]),
    });

    if (cleanXml.error) {
      log.error(cleanXml.error);
      process.exit(1);
    }

    return loadSvg(cleanXml.data, glyph, config.start_codepoint);
  });

  var buildConfig = fontConfig(
    _.extend({}, config, {
      glyphs: glyphs,
    })
  );

  generateFonts(buildConfig);

  return buildConfig;
};
